<a href="https://colab.research.google.com/github/MattosJ/Programas-em-Python/blob/main/Hora%2Bda%2Bpratica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print('Escola de Dados da Alura!')

Escola de Dados da Alura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
print('Nome: Matheus')
print('Sobrenome: Araujo Pereira')

Nome: Matheus
Sobrenome: Araujo Pereira


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [ ]:
print('M')
print('A')
print('T')
print('H')
print('E')
print('U')
print('S')

M
A
T
H
E
U
S


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
print(19, 'junho', 2001)

19 junho 2001


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
print('Ano atual:', 2023)

Ano atual: 2023


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
nome = input('Digite o seu nome: ')
print(f'Olá, {nome}!')

Digite o seu nome: Matheus
Olá, Matheus!


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
nome = input('Digite o seu nome: ')
idade = int(input('Digite a sua idade: '))
print('Olá, %s, você tem %d anos.' %(nome, idade))

Digite o seu nome: Matheus
Digite a sua idade: 21
Olá, Matheus, você tem 21 anos.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
nome = input('Digite o seu nome: ')
idade = int(input('Digite a sua idade: '))
altura = float(input('Digite a sua altura em metros: '))
print('Olá, %s, você tem %d anos e mede %f metros!.' %(nome, idade, altura))

Digite o seu nome: Matheus
Digite a sua idade: 21
Digite a sua altura em metros: 1.78
Olá, Matheus, você tem 21 anos e mede 1.780000 metros!.


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
print(n1 + n2)

Digite o primeiro número: 14
Digite o segundo número: 15
29.0


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
n3 = float(input('Digite o terceiro número: '))
print(n1 + n2 + n3)

Digite o primeiro número: 4.5
Digite o segundo número: 9.3
Digite o terceiro número: 5.5
19.3


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
print(n1 - n2)

Digite o primeiro número: 1
Digite o segundo número: 1
0.0


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
n1 = float(input('Digite o primeiro número: '))
n2 = float(input('Digite o segundo número: '))
print(n1 * n2)

Digite o primeiro número: 2
Digite o segundo número: 4
8.0


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
n1 = float(input('Digite o numerador: '))
n2 = float(input('Digite o denominador, e saiba que ele não pode ser 0: '))
print(n1 / n2)

Digite o numerador: 10
Digite o denominador, e saiba que ele não pode ser 0: 2
5.0


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
n1 = float(input('Digite o operador: '))
n2 = float(input('Digite a potência: '))
print(n1 ** n2)

Digite o operador: 2
Digite a potência: 3
8.0


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
n1 = float(input('Digite o numerador: '))
n2 = float(input('Digite o denominador, e saiba que ele não pode ser 0: '))
print(n1 // n2)

Digite o numerador: 7
Digite o denominador, e saiba que ele não pode ser 0: 3
2.0


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
n1 = float(input('Digite o numerador: '))
n2 = float(input('Digite o denominador, e saiba que ele não pode ser 0: '))
print(n1 % n2)

Digite o numerador: 7
Digite o denominador, e saiba que ele não pode ser 0: 3
1.0


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
n1 = float(input('Digite a nota 1: '))
n2 = float(input('Digite a nota 2: '))
n3 = float(input('Digite a nota 3: '))
print((n1 + n2 + n3) / (3))

Digite a nota 1: 3
Digite a nota 2: 5
Digite a nota 3: 7
5.0


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
n1 = 5
n2 = 12
n3 = 20
n4 = 15
p1 = 1
p2 = 2
p3 = 3
p4 = 4
ponderacao_total = p1 + p2 + p3 + p4
media_ponderada = ((n1 * p1 + n2 * p2 + n3 * p3 + n4 * p4) / (ponderacao_total))
media_ponderada

14.9

### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase = 'Minha frase!'
print(frase)

Minha frase!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frase = input('Digite sua frase: ')
print(frase)

Digite sua frase: Está é a minha frase!
Está é a minha frase!


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.upper()
print(frase)

Digite sua frase: Está é a minha frase!
ESTÁ É A MINHA FRASE!


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.lower()
print(frase)

Digite sua frase: Está é a minha frase!
está é a minha frase!


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase = ' Minha string '
frase = frase.strip()
frase

'Minha string'

#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.strip()
frase

Digite sua frase:      Está é a minha frase!    


'Está é a minha frase!'

#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.lower()
frase

Digite sua frase: Minha frase


'minha frase'

#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.replace('e', 'f')
frase

Digite sua frase: Eu estou teste uma frase de f no lugar de e!


'Eu fstou tfstf uma frasf df f no lugar df f!'

#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.replace('a', '@')
frase

Digite sua frase: Ainda bem que eu sei trocar o @ pelo a!


'Aind@ bem que eu sei troc@r o @ pelo @!'

#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input('Digite sua frase: ')
frase = frase.replace('s', '$')
frase

Digite sua frase: Todas as pessoas gostam ter varias notas de dinheiro!


'Toda$ a$ pe$$oa$ go$tam ter varia$ nota$ de dinheiro!'

## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
n1 = float(input('Digite um número: '))
n2 = float(input('Digite outro número: '))

if n1 > n2 :
  print(n1)
else :
  print(n2)

Digite um número: 10
Digite outro número: 11
11.0


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
percentual = float(input('Digite o percentual de crescimento de produção: '))

if percentual >= 0 :
  print('Crecimento, pois porcentagem positiva!')
else :
  print('Decrescimento, porcentagem negativa!')

Digite o percentual de crescimento de produção: 100
Crecimento, pois porcentagem positiva!


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
letra_fornecida = input('Digite apenas uma letra: ')

if letra_fornecida == 'a' or 'e' or 'i' or 'o' or 'u' :
  print('Vogal!')
else :
  print('Consoante!')

Digite apenas uma letra: o
Vogal!


In [ ]:
letra_fornecida = input('Digite apenas uma letra: ')

if letra_fornecida == 'a' :
  print('Vogal!')
elif letra_fornecida == 'e' :
  print('Vogal!')
elif letra_fornecida == 'i' :
  print('Vogal!')
elif letra_fornecida == 'o' :
  print('Vogal!')
elif letra_fornecida == 'u' :
  print('Vogal!')
else :
  print('Consoante!')

Digite apenas uma letra: o
Vogal!


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
valor_ano1 = float(input('Digite o valor do carro no 1° ano: '))
valor_ano2 = float(input('Digite o valor do carro no 2° ano: '))
valor_ano3 = float(input('Digite o valor do carro no 3° ano: '))

if valor_ano1 >= valor_ano2 and valor_ano1 >= valor_ano3 :
  print(valor_ano1)
elif valor_ano2 >= valor_ano1 and valor_ano2 >= valor_ano3 :
  print(valor_ano2)
else :
  print(valor_ano3)

if valor_ano1 <= valor_ano2 and valor_ano1 <= valor_ano3 :
  print(valor_ano1)
elif valor_ano2 <= valor_ano1 and valor_ano2 <= valor_ano3 :
  print(valor_ano2)
else :
  print(valor_ano3)

Digite o valor do carro no 1° ano: 100000
Digite o valor do carro no 2° ano: 90000
Digite o valor do carro no 3° ano: 80000
100000.0
80000.0


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
valor_produto1 = float(input('Digite o valor do 1° produto: '))
valor_produto2 = float(input('Digite o valor do 2° produto: '))
valor_produto3 = float(input('Digite o valor do 3° produto: '))

if valor_produto1 <= valor_produto2 and valor_produto1 <= valor_produto3 :
  print('O 1° produto é o mais barato pra comprar!')
elif valor_produto2 <= valor_produto1 and valor_produto2 <= valor_produto3 :
  print('O 2° produto é o mais barato pra comprar!')
else :
  print('O 3° produto é o mais barato pra comprar!')

Digite o valor do 1° produto: 10
Digite o valor do 2° produto: 9
Digite o valor do 3° produto: 8
O 3° produto é o mais barato pra comprar!


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
n1 = float(input('Digite o 1° número: '))
n2 = float(input('Digite o 2° número: '))
n3 = float(input('Digite o 3° número: '))

if n1 <= n2 and n1 <= n3 and n2 <= n3 :
  print(f'{n1}, {n2} e {n3}')
if n1 <= n2 and n1 <= n3 and n3 <= n2 :
  print(f'{n1}, {n3} e {n2}')
if n2 <= n1 and n2 <= n3 and n1 <= n3 :
  print(f'{n2}, {n1} e {n3}')
if n2 <= n1 and n2 <= n3 and n3 <= n1 :
  print(f'{n2}, {n3} e {n1}')
if n3 <= n1 and n3 <= n2 and n1 <= n2 :
  print(f'{n3}, {n1} e {n2}')
if n3 <= n1 and n3 <= n2 and n2 <= n1 :
  print(f'{n3}, {n2} e {n1}')

Digite o 1° número: 3
Digite o 2° número: 2
Digite o 3° número: 1
1.0, 2.0 e 3.0


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [ ]:
turno = input('Qual o turno? ')

if turno == 'manhã' :
  print('Bom Dia!')
elif turno == 'tarde' :
  print('Boa Tarde!')
elif turno == 'noite' :
  print('Boa Noite!')
else :
  print('Valor Inválido!')

Qual o turno? manhã
Bom Dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
n = int(input('Digite um número inteiro para verificar se é par ou ímpar: '))

if n % 2 == 0 :
  print('par')
else :
  print('ímpar')

Digite um número inteiro para verificar se é par ou ímpar: 100
par


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
n = input('Digite um número: ')

if '.' in n :
  print('decimal')
else :
  print('inteiro')

Digite um número: 9.8
decimal


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [ ]:
n1 = input('Digite um número: ')
n2 = input('Digite outro número: ')

resposta = input("""

    -----------------------------------------------------
    Qual operação deseja realizar com os 2 números?

    Digite a para verificar se é par ou ímpar
    Digite b para verificar se é positivo ou negativo
    Digite c para verificar se é inteiro ou decimal
    -----------------------------------------------------

    """)

if resposta == 'a' :

  n1 = float(n1)
  n2 = float(n2)

  if n1 % 2 == 0 :
    print(f'{n1} é par')
  if n1 % 2 != 0 :
    print(f'{n1} é impar')

  if n2 % 2 == 0 :
    print(f'{n2} é par')
  if n2 % 2 != 0 :
    print(f'{n2} é impar')

if resposta == 'b' :

  n1 = float(n1)
  n2 = float(n2)

  if n1 >= 0 :
    print('positivo')
  if n1 < 0 :
    print('negativo')

  if n2 >= 0 :
    print('positivo')
  if n2 < 0 :
    print('negativo')

if resposta == 'c' :

  if '.' in n1 :
    print('decimal')
  else :
    print('inteiro')

  if '.' in n2 :
    print('decimal')
  else :
    print('inteiro')

Digite um número: 10.9
Digite outro número: 15.7


    -----------------------------------------------------
    Qual operação deseja realizar com os 2 números?

    Digite a para verificar se é par ou ímpar
    Digite b para verificar se é positivo ou negativo
    Digite c para verificar se é inteiro ou decimal
    -----------------------------------------------------

    c
decimal
decimal


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [ ]:
lado1 = float(input('Digite o valor do lado 1: '))
lado2 = float(input('Digite o valor do lado 2: '))
lado3 = float(input('Digite o valor do lado 3: '))

if ((lado1 + lado2) > lado3) and ((lado1 + lado3) > lado2) and ((lado2 + lado3) > lado1) :

  if lado1 == lado2 == lado3 :
    print('triângulo equilátero')
  elif (lado1 == lado2) or (lado1 == lado3) or ((lado2 == lado3)) :
    print('triângulo isósceles')
  else :
    print('triângulo escaleno')

else :
  print('Não é triângulo!!!')

Digite o valor do lado 1: 5
Digite o valor do lado 2: 5
Digite o valor do lado 3: 4
triângulo isósceles


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [ ]:
E = 1.7
D = 2.0

combustível = input("""

    Deseja abastecer com etanol ou com diesel?

    Digite etanol para abastecer com etanol
    Digite diesel para abastecer com diesel

    """)

if combustível == 'etanol' :

  quantidade_etanol = input('Quantos litros você quer? ')
  quantidade_etanol = float(quantidade_etanol)

  if 0.0 <= quantidade_etanol <= 15.0 :

    valor_pra_pagar = (E * quantidade_etanol) - (E * quantidade_etanol * 0.02)
    print('Valor a ser pago: ', valor_pra_pagar)

  if quantidade_etanol >= 15.0 :

    valor_pra_pagar = (E * quantidade_etanol) - (E * quantidade_etanol * 0.04)
    print('Valor a ser pago: ', valor_pra_pagar)

if combustível == 'diesel' :

  quantidade_diesel = input('Quantos litros você quer? ')
  quantidade_diesel = float(quantidade_diesel)

  if 0.0 <= quantidade_diesel <= 15.0 :

    valor_pra_pagar = (E * quantidade_diesel) - (E * quantidade_diesel * 0.03)
    print('Valor a ser pago: ', valor_pra_pagar)

  if quantidade_diesel >= 15.0 :

    valor_pra_pagar = (E * quantidade_diesel) - (E * quantidade_diesel * 0.05)
    print('Valor a ser pago: ', valor_pra_pagar)



    Deseja abastecer com etanol ou com diesel?
    
    Digite etanol para abastecer com etanol
    Digite diesel para abastecer com diesel

    etanol
Quantos litros você quer? 18
Valor a ser pago:  29.375999999999998


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [ ]:
quantidade_vendas_2022 = float(input('Informe a quantidade de vendas em 2022: '))
quantidade_vendas_2023 = float(input('Informe a quantidade de vendas em 2023: '))

variacao_percentual = ((quantidade_vendas_2023 - quantidade_vendas_2022) / (quantidade_vendas_2022)) * 100

if variacao_percentual > 20 :
  print('Bonificação para o time de vendas!')

if 2 < variacao_percentual <= 20 :
  print('Pequena bonificação para o time de vendas!')

if -10 <= variacao_percentual <= 2 :
  print('Planejamento de políticas de incentivo às vendas!')

if variacao_percentual < -10 :
  print('Corte de gastos!!!')

Informe a quantidade de vendas em 2022: 15000
Informe a quantidade de vendas em 2023: 20000
Bonificação para o time de vendas!


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [ ]:
n1 = int(input('Digite um número: '))
n2 = int(input('Digite outro número: '))

if n1 < n2 :

  while n1 < n2 :

    print(n1)
    n1 += 1

elif n1 > n2 :

  while n1 > n2 :

    print(n1)
    n1 -= 1

else :

  print(n1 or n2)

Digite um número: 10
Digite outro número: 10
10


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [ ]:
colonia_A = 4
colonia_B = 10

taxa_crescimento_A = 0.015
taxa_crescimento_B = 0.03

while colonia_A < colonia_B :

  colonia_A = colonia_A + colonia_A * taxa_crescimento_A
  colonia_B = colonia_B + colonia_B * taxa_crescimento_B

  print(f'Colônia A: {colonia_A} e Colônia B: {colonia_B}')

  if colonia_A >= colonia_B :

    print(f'Colônia A: {colonia_A} e Colônia B: {colonia_B}')
    print('Colônia A ultrapassou ou igualou a colônia B!')

A saída de streaming foi truncada nas últimas 5000 linhas.
Colônia A: 8.821391368796223e+275 e Colônia B: inf
Colônia A: 8.953712239328166e+275 e Colônia B: inf
Colônia A: 9.088017922918089e+275 e Colônia B: inf
Colônia A: 9.22433819176186e+275 e Colônia B: inf
Colônia A: 9.362703264638288e+275 e Colônia B: inf
Colônia A: 9.503143813607862e+275 e Colônia B: inf
Colônia A: 9.64569097081198e+275 e Colônia B: inf
Colônia A: 9.79037633537416e+275 e Colônia B: inf
Colônia A: 9.937231980404773e+275 e Colônia B: inf
Colônia A: 1.0086290460110844e+276 e Colônia B: inf
Colônia A: 1.0237584817012506e+276 e Colônia B: inf
Colônia A: 1.0391148589267694e+276 e Colônia B: inf
Colônia A: 1.0547015818106709e+276 e Colônia B: inf
Colônia A: 1.070522105537831e+276 e Colônia B: inf
Colônia A: 1.0865799371208984e+276 e Colônia B: inf
Colônia A: 1.1028786361777119e+276 e Colônia B: inf
Colônia A: 1.1194218157203775e+276 e Colônia B: inf
Colônia A: 1.136213142956183e+276 e Colônia B: inf
Colônia A: 1.153256

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [ ]:
quantidade_notas = 1

while quantidade_notas <= 15 :

  try :

    nota = float(input('Digite uma nota: '))

    if 0 <= nota <= 5 :

      print('Nota válida!')
      quantidade_notas += 1

    else :

      print('Insira um valor válido!')

  except :

      print('Isso nem é uma nota amigo!')

Digite uma nota: a
Isso nem é uma nota amigo!
Digite uma nota: 8
Insira um valor válido!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!
Digite uma nota: 1
Nota válida!


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [ ]:
total_temperaturas = 0
quantidade_temperaturas = 0

while quantidade_temperaturas >= 0 :

  temperatura = float(input("""

  Digite o valor da temperatura em Celsius
  Obs.: Para finalizar a leitura e calcular a média de todas as temperaturas escritas digite -273

  """))

  if temperatura != -273 :

    total_temperaturas = total_temperaturas + temperatura
    quantidade_temperaturas += 1

  else :

    if quantidade_temperaturas > 0 :

      print(f'A média de temperaturas é {(total_temperaturas / quantidade_temperaturas)}!!!')
      break

    else :

      break



  Digite o valor da temperatura em Celsius
  Obs.: Para finalizar a leitura e calcular a média de todas as temperaturas escritas digite -273

  5


  Digite o valor da temperatura em Celsius
  Obs.: Para finalizar a leitura e calcular a média de todas as temperaturas escritas digite -273

  10


  Digite o valor da temperatura em Celsius
  Obs.: Para finalizar a leitura e calcular a média de todas as temperaturas escritas digite -273

  -273
A média de temperaturas é 7.5!!!


#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [ ]:
número = int(input('Digite um número para calcular o fatorial: '))
fatorial = 1

for número in range (1, número + 1) :

  fatorial *= número

print(fatorial)

Digite um número para calcular o fatorial: 5
120


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [ ]:
tabuada = int(input('Digite um número para fazermos a tabuada: '))

for número in range(10) :
  print(f'{tabuada} x {número + 1} = {tabuada * (número + 1)}')

Digite um número para fazermos a tabuada: 10
10 x 1 = 10
10 x 2 = 20
10 x 3 = 30
10 x 4 = 40
10 x 5 = 50
10 x 6 = 60
10 x 7 = 70
10 x 8 = 80
10 x 9 = 90
10 x 10 = 100


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [ ]:
def is_prime(n):
    if n <= 1:
        return False

    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return False

    return True

num = int(input("Digite um número inteiro: "))

if is_prime(num):
    print(f"{num} é um número primo.")
else:
    print(f"{num} não é um número primo.")

Digite um número inteiro: 2
2 é um número primo.


#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [ ]:
faixa_1 = faixa_2 = faixa_3 = faixa_4 = 0

while True:
    idade = int(input("Digite a idade do cliente (ou um número negativo para encerrar): "))

    if idade < 0:
        break

    if idade <= 25:
        faixa_1 += 1
    elif idade <= 50:
        faixa_2 += 1
    elif idade <= 75:
        faixa_3 += 1
    else:
        faixa_4 += 1

print(f"Faixa [0-25]: {faixa_1} clientes")
print(f"Faixa [26-50]: {faixa_2} clientes")
print(f"Faixa [51-75]: {faixa_3} clientes")
print(f"Faixa [76-100]: {faixa_4} clientes")


Digite a idade do cliente (ou um número negativo para encerrar): 10
Digite a idade do cliente (ou um número negativo para encerrar): 90
Digite a idade do cliente (ou um número negativo para encerrar): 99
Digite a idade do cliente (ou um número negativo para encerrar): 100
Digite a idade do cliente (ou um número negativo para encerrar): 14
Digite a idade do cliente (ou um número negativo para encerrar): 12
Digite a idade do cliente (ou um número negativo para encerrar): 54
Digite a idade do cliente (ou um número negativo para encerrar): 54
Digite a idade do cliente (ou um número negativo para encerrar): 54
Digite a idade do cliente (ou um número negativo para encerrar): 33
Digite a idade do cliente (ou um número negativo para encerrar): 23
Digite a idade do cliente (ou um número negativo para encerrar): 45
Digite a idade do cliente (ou um número negativo para encerrar): 65
Digite a idade do cliente (ou um número negativo para encerrar): -1
Faixa [0-25]: 4 clientes
Faixa [26-50]: 2 clien

#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [ ]:
candidato_1 = candidato_2 = candidato_3 = candidato_4 = votos_nulos = votos_brancos = 0

for _ in range(20):
    voto = int(input("Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: "))

    if voto == 1:
        candidato_1 += 1
    elif voto == 2:
        candidato_2 += 1
    elif voto == 3:
        candidato_3 += 1
    elif voto == 4:
        candidato_4 += 1
    elif voto == 5:
        votos_nulos += 1
    elif voto == 6:
        votos_brancos += 1

total_votos = candidato_1 + candidato_2 + candidato_3 + candidato_4 + votos_nulos + votos_brancos

print(f"Total de votos para o Candidato 1: {candidato_1}")
print(f"Total de votos para o Candidato 2: {candidato_2}")
print(f"Total de votos para o Candidato 3: {candidato_3}")
print(f"Total de votos para o Candidato 4: {candidato_4}")

print(f"Total de votos nulos: {votos_nulos}")
print(f"Total de votos em branco: {votos_brancos}")

porcentagem_nulos = (votos_nulos / total_votos) * 100
porcentagem_brancos = (votos_brancos / total_votos) * 100

print(f"Porcentagem de votos nulos: {porcentagem_nulos:.2f}%")
print(f"Porcentagem de votos em branco: {porcentagem_brancos:.2f}%")


Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 1
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 2
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 3
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 4
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 1
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 2
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 3
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 4
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 1
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 2
Digite o número do candidato (1, 2, 3, 4), 5 para voto nulo ou 6 para voto em branco: 3
Digite o número do candidato (1,

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

In [ ]:
valor_gastos = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
media = sum(valor_gastos) / len(valor_gastos)
media

3144.222

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

In [ ]:
valor_gastos = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
conte = 0
acima_3000 = []

for valor in valor_gastos :

  if valor > 3000 :

    acima_3000.append(valor)
    conte += 1

print(conte)
divisao = (sum(acima_3000) / sum(valor_gastos)) * 100
divisao

7


76.68421631805896

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

In [ ]:
numero = input('Digite um número: ')

Digite um número: [1,2,3,4,5]


#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

In [ ]:
numero1 = input('Digite um número: ')
numero2 = input('Digite um número: ')
numero3 = input('Digite um número: ')
numero4 = input('Digite um número: ')
numero5 = input('Digite um número: ')

lista = []

lista.append(numero5)
lista.append(numero4)
lista.append(numero3)
lista.append(numero2)
lista.append(numero1)

print(lista)

Digite um número: 1
Digite um número: 2
Digite um número: 3
Digite um número: 4
Digite um número: 5
['5', '4', '3', '2', '1']


#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

In [ ]:
numero = int(input("Digite um número: "))

if numero >= 2:
    primos = []
    for i in range(2, numero + 1):
        eh_primo = True
        for j in range(2, i):
            if i % j == 0:
                eh_primo = False
                break
        if eh_primo:
            primos.append(i)
    print("Números primos entre 1 e", numero, ":", primos)
else:
    print("O número deve ser maior ou igual a 2.")


Digite um número: 15
Números primos entre 1 e 15 : [2, 3, 5, 7, 11, 13]


#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

In [ ]:
dia = int(input("Digite o dia: "))
mes = int(input("Digite o mês: "))
ano = int(input("Digite o ano: "))

data_valida = True

if mes < 1 or mes > 12:
    data_valida = False
elif dia < 1:
    data_valida = False
elif mes == 2:
    if dia > 28:
        data_valida = False
elif mes in [4, 6, 9, 11]:
    if dia > 30:
        data_valida = False
else:
    if dia > 31:
        data_valida = False

if ano < 1:
    data_valida = False

if data_valida:
    print("A data", dia, "/", mes, "/", ano, "é válida para análise.")
else:
    print("A data", dia, "/", mes, "/", ano, "não é válida para análise.")


Digite o dia: 40
Digite o mês: 10
Digite o ano: 2001
A data 40 / 10 / 2001 não é válida para análise.


### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
